external/fv3fit/fv3fit/sklearn/_min_max_novelty_detector.py

```diff
@@ -0,0 +1,131 @@
+from fv3fit._shared import stacking
+from .. import _shared
+from .._shared import (
+    match_prediction_to_input_coords,
+    pack,
+    pack_tfdataset,
+    Predictor,
+    register_training_function,
+    SAMPLE_DIM_NAME,
+    stack_non_vertical,
+)
+from .._shared.config import MinMaxNoveltyDetectorHyperparameters, PackerConfig
+
+import dataclasses
+import dacite
+import fsspec
+from fv3fit import tfdataset
+from fv3fit._shared.packer import PackingInfo, clip_sample, unpack
+from fv3fit.tfdataset import apply_to_mapping, ensure_nd
+from fv3fit.typing import Batch
+import io
+import joblib
+import logging
+import numpy as np
+from sklearn.preprocessing import MinMaxScaler
+import tensorflow as tf
+from typing import Optional, Iterable, Sequence
+from vcm import safe
+import xarray as xr
+import yaml
+
+@register_training_function("min_max_novelty_detector", MinMaxNoveltyDetectorHyperparameters)
+def train_min_max_novelty_detector(
+    hyperparameters: MinMaxNoveltyDetectorHyperparameters,
+    train_batches: tf.data.Dataset,
+    validation_batches: tf.data.Dataset
+):
+    train_batches = train_batches.map(
+        tfdataset.apply_to_mapping(tfdataset.float64_to_float32)
+    )
+    
+    model = MinMaxNoveltyDetector(hyperparameters)
+    model.fit(train_batches)
+    return model
+
+@_shared.io.register("minmax")
+class MinMaxNoveltyDetector(Predictor):
+
+    _PICKLE_NAME = "minmax.pkl"
+    _METADATA_NAME = "metadata.bin"
+
+    _NOVELTY_OUTPUT_VAR = "is_novelty"
+
+    dims = ["x", "y", "tile", "time"]
+
+    def __init__(
+        self,
+        hyperparameters: MinMaxNoveltyDetectorHyperparameters
+    ):
+        output_variables = [self._NOVELTY_OUTPUT_VAR]
+        super().__init__(hyperparameters.input_variables, output_variables)
+        self.packer_config = hyperparameters.packer_config
+
+    def fit(self, batches: tf.data.Dataset):
```

Comment:
Feel free to ignore this, but just an FYI from previous iterations of code like this.

Having `fit` as a method on the predictor is something we've had on our predictors before, and makes the APIs a bit more complicated. For example, it's possible to call `predict` before `fit`, which means you either need error message logic or you'll give a confusing error message, and it's possible to call `fit` after `load` or `dump` before `fit`, etc. It's also possible to call `fit` multiple times, in which case it's not obvious what should happen (should the training be re-done? updated incrementally?).

These issues are avoided if you make a `fit(batches) -> Predictor` function. For example, if you make `input_features_` and `scaler` arguments of `__init__` for this class, you don't need a fit method because the fitting happens before constructing the predictor.

Reply:
I'll make this change once I get back; it should be easy. My only hesitation was that `predict` and `fit` the way they're currently defined are the way things are done with `sklearn` classes (albeit with exceptions if you call predict on an untrained model) and how `_random_forest.py` is configured too. Would you like me to change random forest to make it too match this convention?

Reply:
Refactoring the random forest class to behave like this has been on my to-do list for almost a year, just isn't high enough priority since the code works and we haven't needed to make any changes to it in that time. I wouldn't ask you to do it since you only have so much time with us, but if you want to do this refactor ask your mentors so they can help you prioritize. If you refactor the RF class I'd suggest doing it in a different PR to separate the changes.